Open Judge System/Web/OJS.Web/Areas/Administration/Controllers/ContestsController.cs

```diff
@@ -375,43 +368,79 @@ public ActionResult ChangeActiveParticipantsEndTime(ChangeTimeForParticipantsVie
+                {
+                    this.ModelState.AddModelError(
+                        nameof(model.ParticipantsCreatedAfterDateTime),
+                        ChangeTimeResource.Participants_created_after_must_be_before_Participants_created_before);
+                }
+                else
+                {
+                    isModelValid = true;
+                }
 
-            var successMessage = model.TimeInMinutes >= 0
-                ? string.Format(Resource.Added_time_to_participants_online, minutesForDisplay, model.ContestName)
-                : string.Format(
-                    Resource.Subtracted_time_from_participants_online,
-                    minutesForDisplay.Substring(1, minutesForDisplay.Length - 1),
-                    model.ContestName);
-            sb.AppendLine(successMessage);
+                if (!isModelValid)
+                {
+                    PrepareViewBagData();
+                    return this.View(model);
+                }
 
-            if (notUpdatedUsersUsernames.Any())
+                result = this.participantsBusiness
+                    .UpdateParticipationsEndTimeByContestByParticipationStartTimeRangeAndTimeInMinutes(
+                        model.ContesId,
+                        model.TimeInMinutes,
+                        model.ParticipantsCreatedAfterDateTime.Value,
+                        model.ParticipantsCreatedBeforeDateTime.Value);
+            }
+            else
             {
-                var warningMessage = string.Format(
-                    Resource.Failed_to_update_participants_duration,
-                    string.Join(", ", notUpdatedUsersUsernames));
-              sb.AppendLine(warningMessage);
+                var participant = this.participantsData
+                    .GetByContestByUserAndByIsOfficial(model.ContesId, model.UserId, true);
+
+                if (participant == null)
+                {
+                    this.ModelState.AddModelError(nameof(model.UserId), Resource.Participant_not_in_contest);
+                    PrepareViewBagData();
+                    return this.View(model);
+                }
+
+                result = this.participantsBusiness
+                    .UpdateParticipationEndTimeByIdAndTimeInMinutes(participant.Id, model.TimeInMinutes);
+            }
+
+            if (!result.IsError)
+            {
+                var systemMessage = this.GetMessageForChangedParticipantsEndTime(
+                    model.TimeInMinutes,
+                    model.ContestName,
+                    result);
+
+                this.TempData.AddInfoMessage(systemMessage);
+            }
+            else
+            {
+                this.TempData.AddDangerMessage(result.Error);
             }
 
-            this.TempData.AddInfoMessage(sb.ToString());
+            void PrepareViewBagData() => this.ViewBag.CurrentUsername = this.UserProfile.UserName;
```

Comment:
What is that syntax? I haven't seen it before. Is this a short-hand for a private method? If so why not define it in the controller body, as opposed to inside the method body. Interesting.

Reply:
This ViewBag data is specific for this action and it has to be set every time before returning the view, which is done several times in the body of the method and a [local function](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/local-functions) seems like an appropriate solution for not repeating the code.